src/HttpServer.gren

```diff
@@ -0,0 +1,453 @@
+                        , command = Stream.sendLine model.stderr <|
+                            "Server failed to start: " ++ code ++ "\n" ++ message
+                        }
+        
+            GotRequest req res ->
+                let
+                    body = "You requested: " ++ Url.toString req.url
+                in
+                { model = model
+                , command =
+                    res
+                        |> Response.setHeader "Content-type" "text/html"
+                        |> Response.setBody ("<html>" ++ body ++ "</html>")
+                        |> Response.send
+                }
+
+    subscriptions : Model -> Sub Msg
+    subscriptions model =
+        case model.server of
+            Just server ->
+                Http.onRequest server GotRequest
+
+            Nothing ->
+                Sub.none
+
+See [examples/http-server](https://github.com/gren-lang/node/tree/main/examples/http-server) for more.
+
+## Initialization
+
+@docs Permission, Server, ServerError, initialize, createServer
+
+## Requests
+
+@docs Request, Method, bodyAsString, bodyFromJson, toRequest, requestInfo
+
+## Responding to requests
+
+@docs onRequest
+
+See [HttpServer.Response](HttpServer.Response) for more details on responding to requests.
+
+-}
+
+import Bytes exposing (Bytes)
+import Bytes.Decode as Decode
+import Dict exposing (Dict)
+import Init
+import Internal.Init
+import Json.Encode
+import Json.Decode
+import Node
+import Task exposing (Task)
+import Gren.Kernel.HttpServer
+import HttpServer.Response exposing (Response(..))
+import Url exposing (Url, Protocol(..))
+
+
+-- INITIALIZATION
+
+
+{-| The permission to start a [`Server`](HttpServer.Server).
+
+You get this from [`initialize`](HttpServer.initialize).
+-}
+type Permission
+    = Permission
+
+{-| The HTTP server.
+-}
+type Server
+    -- Note: Actual implementation in Kernel code
+    = Server
+
+
+{-| Error code and message from node.
+Most likely from a failed attempt to start the server (e.g. `EADDRINUSE`).
+Refer to the [node docs](https://nodejs.org/docs/latest-v18.x/api/errors.html) for details.
+-}
+type ServerError =
+    ServerError String String
+
+
+{-| Initialize the [`HttpServer`](HttpServer) module and get permission to create a server.
+-}
+initialize : Init.Task Permission
+initialize =
+    Task.succeed Permission
+        |> Internal.Init.Task
+
+
+
+{-| Task to initialize a [`Server`](HttpServer#Server).
+-}
+createServer : Permission -> { host : String, port_ : Int } -> Task ServerError Server
+createServer _ options =
+    Gren.Kernel.HttpServer.createServer options.host options.port_
+
+
+-- REQUESTS
+
+
+{-| An incoming HTTP reqest.
+-}
+type alias Request =
+    { headers : Dict String String
+    , method : Method
+    , body : Bytes
+    , url : Url
+    }
+
+
+{-| HTTP request methods.
+-}
+type Method
+    = GET
+    | HEAD
+    | POST
+    | PUT
+    | DELETE
+    | CONNECT
+    | TRACE
+    | PATCH
+    | UNKNOWN
```

Comment:
Should  `UNKNOWN` contain the string it failed to recognize?

Reply:
Yep! Unfortunately in practice this will result in an unhandled 400 error because node won't accept custom request methods.

Reply:
If I'm understanding you correctly: `UNKNOWN` won't ever be created in practice because `node` blocks such requests?

In that case, could we leave out the `UNKNOWN` constructor entirely?

Reply:
That's correct. But if we leave it off, `toMethod` would have to choose a real http method as the default [when we turn the incoming method string from node into a gren type](https://github.com/gren-lang/node/pull/8/files#diff-b6d58b86d734deb492ee4940ea3eb6f1ebe84b5d1ef1e484019b1733ecc4e054R50). Wouldn't be a problem in practice I guess, but I like that what's written now is forward compatible with eventual support for non-standard methods. Not sure what the safest path is if we don't want some non-standard fallback here, besides letting `Request.method` be a string in gren?

Reply:
Fair enough. We'll leave it as is, then :)